Add a GKE cleanup workflow to run once per hour.

zcain117 opened the pull request:
## What does this PR do?

Once per hour, kill and delete all GKE jobs that have an age >1hr. This prevents buildup of stale jobs and prevents and erroneous long-running jobs in case the job accidentally has a long timeout.

Fixes # (issue)

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

<!-- For CHANGELOG separate each item in unreleased section by a blank line to reduce collisions -->

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

zcain117 commented:
I'm not sure the best way to test this. I can temporarily make the `job` run for incoming commits, see at least the job finishes and succeeds, then turn off the job running for commits and go back to `schedule`.

Or maybe CircleCI will run `schedule` workflows even for pending PRs?  I guess we will find out in 58 minutes

Borda commented on .circleci/config.yml:
```diff
@@ -123,9 +133,32 @@ jobs:
+  cleanup-gke-jobs:
+    docker:
+      - image: circleci/python:3.7
+    environment:
+      - MAX_CHECKS: 60
```
```suggestion
```
not needed here

Borda commented on .circleci/config.yml:
```diff
@@ -123,9 +133,32 @@ jobs:
+  cleanup:
+    triggers:
+      - schedule:
+          cron: "0 * * * *"
```
```suggestion
          # The cron format is:
          # min (0-59) hour (0-23) monthday (1-31) month (1-12) weekday (0-6, 0=Sun)
          cron: "0 * * * *"
```

Borda commented on .circleci/config.yml:
```diff
@@ -85,6 +85,16 @@ references:
        sudo pip install pycobertura
        pycobertura show coverage.xml
 
+  delete_jobs: &delete_jobs
```
```suggestion
  delete_gke_jobs: &delete_gke_jobs
```

Borda commented:
> I'm not sure the best way to test this. I can temporarily make the `job` run for incoming commits, see at least the job finishes and succeeds, then turn off the job running for commits and go back to `schedule`.
> 
> Or maybe CircleCI will run `schedule` workflows even for pending PRs? I guess we will find out in 58 minutes



> I'm not sure the best way to test this. I can temporarily make the `job` run for incoming commits, see at least the job finishes and succeeds, then turn off the job running for commits and go back to `schedule`.
> 
> Or maybe CircleCI will run `schedule` workflows even for pending PRs? I guess we will find out in 58 minutes

not sure if I follow here... by my understanding corn shall run independently on any open PR...

for testing, we can make a million loop on running tests (plus some sleep in between repetitions) for a while [just in case make the PR in this time as "wip"]

zcain117 commented:
I changed the `job` temporarily to run on incoming commit and got a successful CI run [here](https://circleci.com/gh/PyTorchLightning/pytorch-lightning/49318?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)

Now I switched it back to `trigger` / `schedule` so that it will run every hour once submitted and won't run for every PR / commit
